File: nodejs-compressor/src/lexer/Lexer.js
/**
 * @fileoverview Lexer for ASON 2.0 format
 *
 * Tokenizes ASON 2.0 text into a stream of tokens for the parser.
 * Handles all ASON 2.0 syntax including sections, arrays, references, and values.
 *
 * @module Lexer
 * @license MIT
 * @version 2.0.0
 */

import { Token } from './Token.js';
import { TokenType } from './TokenType.js';

/**
 * Lexical analyzer (tokenizer) for ASON 2.0 format.
 *
 * Converts raw text into a stream of tokens that can be parsed.
 * Maintains position tracking for error reporting.
 *
 * @class Lexer
 *
 * @example
 * const lexer = new Lexer('@customer\n  name:John');
 * const tokens = lexer.tokenize();
 */
export class Lexer {
  /**
   * Creates a new Lexer instance.
   *
   * @constructor
   * @param {string} input - ASON 2.0 text to tokenize
   */
  constructor(input) {
    /** @type {string} Raw input text */
    this.input = input;

    /** @type {number} Current position in input (0-indexed) */
    this.pos = 0;

    /** @type {number} Current line number (1-indexed) */
    this.line = 1;

    /** @type {number} Current column number (1-indexed) */
    this.column = 1;

    /** @type {Token[]} Accumulated tokens */
    this.tokens = [];
  }

  /**
   * Gets the current character without consuming it.
   *
   * @returns {string|null} Current character or null if at end
   */
  peek() {
    return this.pos < this.input.length ? this.input[this.pos] : null;
  }

  /**
   * Gets a character at offset from current position.
   *
   * @param {number} offset - Offset from current position
   * @returns {string|null} Character at offset or null
   */
  peekAt(offset) {
    const targetPos = this.pos + offset;
    return targetPos < this.input.length ? this.input[targetPos] : null;
  }

  /**
   * Consumes and returns the current character, advancing position.
   *
   * @returns {string|null} Current character or null if at end
   */
  advance() {
    if (this.pos >= this.input.length) return null;

    const char = this.input[this.pos];
    this.pos++;

    if (char === '\n') {
      this.line++;
      this.column = 1;
    } else {
      this.column++;
    }

    return char;
  }

  /**
   * Checks if current position is at end of input.
   *
   * @returns {boolean} True if at end of input
   */
  isAtEnd() {
    return this.pos >= this.input.length;
  }

  /**
   * Skips whitespace characters (spaces and tabs only, not newlines).
   *
   * @returns {number} Number of spaces skipped
   */
  skipWhitespace() {
    let count = 0;
    while (this.peek() === ' ' || this.peek() === '\t') {
      this.advance();
      count++;
    }
    return count;
  }

  /**
   * Checks if a character is a valid identifier start character.
   *
   * @param {string} char - Character to check
   * @returns {boolean} True if valid identifier start
   */
  isIdentifierStart(char) {
    if (!char) return false;
    return /[a-zA-Z_]/.test(char);
  }

  /**
   * Checks if a character is a valid identifier character.
   *
   * @param {string} char - Character to check
   * @returns {boolean} True if valid identifier character
   */
  isIdentifierChar(char) {
    if (!char) return false;
    return /[a-zA-Z0-9_]/.test(char);
  }

  /**
   * Checks if a character is a digit.
   *
   * @param {string} char - Character to check
   * @returns {boolean} True if digit
   */
  isDigit(char) {
    if (!char) return false;
    return /[0-9]/.test(char);
  }

  /**
   * Tokenizes the entire input.
   *
   * @returns {Token[]} Array of tokens
   */
  tokenize() {
    this.tokens = [];

    while (!this.isAtEnd()) {
      this.tokenizeNext();
    }

    // Add EOF token
    this.tokens.push(Token.eof(this.line, this.column));

    return this.tokens;
  }

  /**
   * Tokenizes the next token from current position.
   */
  tokenizeNext() {
    const char = this.peek();
    const startLine = this.line;
    const startColumn = this.column;

    // Newline
    if (char === '\n' || char === '\r') {
      this.tokenizeNewline();
      return;
    }

    // Whitespace (indentation at start of line)
    if (char === ' ' || char === '\t') {
      // Only create INDENT token at start of line
      if (this.column === 1 || this.tokens.length === 0 ||
          this.tokens[this.tokens.length - 1].type === TokenType.NEWLINE) {
        const spaces = this.skipWhitespace();
        if (spaces > 0) {
          this.tokens.push(Token.indent(spaces, startLine, startColumn));
        }
      } else {
        // Skip inline whitespace
        this.skipWhitespace();
      }
      return;
    }

    // Comment
    if (char === '#') {
      this.tokenizeComment();
      return;
    }

    // Section marker
    if (char === '@') {
      this.tokenizeSection();
      return;
    }

    // Colon
    if (char === ':') {
      this.advance();
      this.tokens.push(new Token(TokenType.COLON, ':', startLine, startColumn, 1));
      return;
    }

    // Pipe
    if (char === '|') {
      this.advance();
      this.tokens.push(new Token(TokenType.PIPE, '|', startLine, startColumn, 1));
      return;
    }

    // Number (including negative) - Check BEFORE dash to handle negative numbers
    if (this.isDigit(char) || (char === '-' && this.isDigit(this.peekAt(1)))) {
      this.tokenizeNumber();
      return;
    }

    // Dash (array item) - Only if not followed by digit (negative numbers handled above)
    if (char === '-') {
      this.advance();
      this.tokens.push(new Token(TokenType.DASH, '-', startLine, startColumn, 1));
      return;
    }

    // Left brace
    if (char === '{') {
      this.advance();
      this.tokens.push(new Token(TokenType.LBRACE, '{', startLine, startColumn, 1));
      return;
    }

    // Right brace
    if (char === '}') {
      this.advance();
      this.tokens.push(new Token(TokenType.RBRACE, '}', startLine, startColumn, 1));
      return;
    }

    // Left bracket
    if (char === '[') {
      this.advance();
      this.tokens.push(new Token(TokenType.LBRACKET, '[', startLine, startColumn, 1));
      return;
    }

    // Right bracket
    if (char === ']') {
      this.advance();
      this.tokens.push(new Token(TokenType.RBRACKET, ']', startLine, startColumn, 1));
      return;
    }

    // Comma
    if (char === ',') {
      this.advance();
      this.tokens.push(new Token(TokenType.COMMA, ',', startLine, startColumn, 1));
      return;
    }

    // Dot
    if (char === '.') {
      this.advance();
      this.tokens.push(new Token(TokenType.DOT, '.', startLine, startColumn, 1));
      return;
    }

    // Variable reference ($var)
    if (char === '$') {
      this.tokenizeReference('$', TokenType.VAR_REF);
      return;
    }

    // Object reference (&obj)
    if (char === '&') {
      this.tokenizeReference('&', TokenType.OBJ_REF);
      return;
    }

    // Quoted string
    if (char === '"' || char === "'") {
      this.tokenizeQuotedString(char);
      return;
    }

    // Identifier, keyword, or unquoted string
    if (this.isIdentifierStart(char)) {
      this.tokenizeIdentifierOrKeyword();
      return;
    }

    // Unknown character - skip it
    this.advance();
  }

  /**
   * Tokenizes a newline (handles \n, \r\n, \r).
   */
  tokenizeNewline() {
    const startLine = this.line;
    const startColumn = this.column;
    let value = '';

    if (this.peek() === '\r' && this.peekAt(1) === '\n') {
      value = this.advance() + this.advance();
    } else {
      value = this.advance();
    }

    this.tokens.push(new Token(TokenType.NEWLINE, value, startLine, startColumn, value.length));
  }

  /**
   * Tokenizes a comment (# or #| |#).
   */
  tokenizeComment() {
    const startLine = this.line;
    const startColumn = this.column;

    this.advance(); // consume #

    // Multi-line comment start #|
    if (this.peek() === '|') {
      this.advance(); // consume |
      this.tokens.push(new Token(TokenType.COMMENT_START, '#|', startLine, startColumn, 2));
      return;
    }

    // Single-line comment - consume until newline
    let value = '#';
    while (!this.isAtEnd() && this.peek() !== '\n' && this.peek() !== '\r') {
      value += this.advance();
    }

    this.tokens.push(new Token(TokenType.COMMENT, value, startLine, startColumn, value.length));
  }

  /**
   * Tokenizes a section marker (@section_name).
   */
  tokenizeSection() {
    const startLine = this.line;
    const startColumn = this.column;

    this.advance(); // consume @

    // Read section name (identifier)
    let name = '';
    while (this.isIdentifierChar(this.peek()) || this.peek() === '.') {
      name += this.advance();
    }

    const value = '@' + name;
    this.tokens.push(new Token(TokenType.SECTION, value, startLine, startColumn, value.length));
  }

  /**
   * Tokenizes a reference ($var or &obj).
   *
   * @param {string} prefix - Reference prefix ($ or &)
   * @param {string} type - Token type
   */
  tokenizeReference(prefix, type) {
    const startLine = this.line;
    const startColumn = this.column;

    this.advance(); // consume prefix

    // Check for keywords like $def: or $data:
    if (prefix === '$') {
      const remaining = this.input.substring(this.pos, this.pos + 5);
      if (remaining.startsWith('def:')) {
        this.advance(); this.advance(); this.advance(); this.advance(); // consume 'def:'
        this.tokens.push(new Token(TokenType.DEF_KEYWORD, '$def:', startLine, startColumn, 5));
        return;
      }
      if (remaining.startsWith('data:')) {
        this.advance(); this.advance(); this.advance(); this.advance(); this.advance(); // consume 'data:'
        this.tokens.push(new Token(TokenType.DATA_KEYWORD, '$data:', startLine, startColumn, 6));
        return;
      }
    }

    // Read reference name
    let name = '';
    while (this.isIdentifierChar(this.peek())) {
      name += this.advance();
    }

    const value = prefix + name;
    this.tokens.push(new Token(type, value, startLine, startColumn, value.length));
  }

  /**
   * Tokenizes a quoted string (supports " and ').
   *
   * @param {string} quote - Quote character (" or ')
   */
  tokenizeQuotedString(quote) {
    const startLine = this.line;
    const startColumn = this.column;

    this.advance(); // consume opening quote

    let value = '';
    let escaped = false;

    while (!this.isAtEnd()) {
      const char = this.peek();

      if (escaped) {
        value += this.advance();
        escaped = false;
      } else if (char === '\\') {
        value += this.advance();
        escaped = true;
      } else if (char === quote) {
        this.advance(); // consume closing quote
        break;
      } else {
        value += this.advance();
      }
    }

    // Return the value WITH quotes so parser can distinguish quoted vs unquoted
    const fullValue = quote + value + quote;
    this.tokens.push(new Token(TokenType.STRING, fullValue, startLine, startColumn, fullValue.length));
  }

  /**
   * Tokenizes a number (integer or float, including negative).
   */
  tokenizeNumber() {
    const startLine = this.line;
    const startColumn = this.column;

    let value = '';

    // Handle negative sign
    if (this.peek() === '-') {
      value += this.advance();
    }

    // Read digits
    while (this.isDigit(this.peek())) {
      value += this.advance();
    }

    // Read decimal part
    if (this.peek() === '.' && this.isDigit(this.peekAt(1))) {
      value += this.advance(); // consume .
      while (this.isDigit(this.peek())) {
        value += this.advance();
      }
    }

    // Read scientific notation (e.g., 1.5e10, 2e-3)
    // Only consume 'e'/'E' if followed by digits (optionally preceded by +/-)
    if (this.peek() === 'e' || this.peek() === 'E') {
      const nextChar = this.peekAt(1);
      const hasSign = nextChar === '+' || nextChar === '-';
      const charAfterSign = hasSign ? this.peekAt(2) : nextChar;

      // Only treat as scientific notation if there's a digit after e/E (and optional sign)
      if (this.isDigit(charAfterSign)) {
        value += this.advance(); // consume 'e' or 'E'
        if (hasSign) {
          value += this.advance(); // consume '+' or '-'
        }
        while (this.isDigit(this.peek())) {
          value += this.advance();
        }
      }
    }

    this.tokens.push(new Token(TokenType.NUMBER, value, startLine, startColumn, value.length));
  }

  /**
   * Tokenizes an identifier, keyword, or unquoted string value.
   */
  tokenizeIdentifierOrKeyword() {
    const startLine = this.line;
    const startColumn = this.column;

    let value = '';

    // Read identifier characters
    while (this.isIdentifierChar(this.peek()) || this.peek() === '.') {
      value += this.advance();
    }

    // Check for keywords (true, false, null)
    if (value === 'true' || value === 'false') {
      this.tokens.push(new Token(TokenType.BOOLEAN, value, startLine, startColumn, value.length));
      return;
    }

    if (value === 'null') {
      this.tokens.push(new Token(TokenType.NULL, value, startLine, startColumn, value.length));
      return;
    }

    // Otherwise it's an identifier or unquoted string
    this.tokens.push(new Token(TokenType.IDENTIFIER, value, startLine, startColumn, value.length));
  }

  /**
   * Filters out ignorable tokens (comments, whitespace).
   *
   * @param {Token[]} tokens - Tokens to filter
   * @returns {Token[]} Filtered tokens
   */
  static filterIgnorable(tokens) {
    return tokens.filter(token =>
      token.type !== TokenType.COMMENT &&
      token.type !== TokenType.COMMENT_START &&
      token.type !== TokenType.COMMENT_END
    );
  }
}

File: python-compressor/src/ason/lexer/lexer.py
"""
Lexer for ASON 2.0 format

Tokenizes ASON 2.0 text into a stream of tokens for the parser.
Handles all ASON 2.0 syntax including sections, arrays, references, and values.

Module: lexer
License: MIT
Version: 2.0.0
"""

import re
from typing import List, Optional
from .token import Token
from .token_type import TokenType


class Lexer:
    """
    Lexical analyzer (tokenizer) for ASON 2.0 format.

    Converts raw text into a stream of tokens that can be parsed.
    Maintains position tracking for error reporting.

    Example:
        >>> lexer = Lexer('@customer\\n  name:John')
        >>> tokens = lexer.tokenize()
    """

    def __init__(self, input_text: str):
        """
        Create a new Lexer instance.

        Args:
            input_text: ASON 2.0 text to tokenize
        """
        self.input = input_text
        self.pos = 0
        self.line = 1
        self.column = 1
        self.tokens: List[Token] = []

    def peek(self) -> Optional[str]:
        """
        Get the current character without consuming it.

        Returns:
            Current character or None if at end
        """
        return self.input[self.pos] if self.pos < len(self.input) else None

    def peek_at(self, offset: int) -> Optional[str]:
        """
        Get a character at offset from current position.

        Args:
            offset: Offset from current position

        Returns:
            Character at offset or None
        """
        target_pos = self.pos + offset
        return self.input[target_pos] if target_pos < len(self.input) else None

    def advance(self) -> Optional[str]:
        """
        Consume and return the current character, advancing position.

        Returns:
            Current character or None if at end
        """
        if self.pos >= len(self.input):
            return None

        char = self.input[self.pos]
        self.pos += 1

        if char == '\n':
            self.line += 1
            self.column = 1
        else:
            self.column += 1

        return char

    def is_at_end(self) -> bool:
        """
        Check if current position is at end of input.

        Returns:
            True if at end of input
        """
        return self.pos >= len(self.input)

    def skip_whitespace(self) -> int:
        """
        Skip whitespace characters (spaces and tabs only, not newlines).

        Returns:
            Number of spaces skipped
        """
        count = 0
        while self.peek() in [' ', '\t']:
            self.advance()
            count += 1
        return count

    def is_identifier_start(self, char: Optional[str]) -> bool:
        """
        Check if a character is a valid identifier start character.

        Args:
            char: Character to check

        Returns:
            True if valid identifier start
        """
        if not char:
            return False
        return bool(re.match(r'[a-zA-Z_]', char))

    def is_identifier_char(self, char: Optional[str]) -> bool:
        """
        Check if a character is a valid identifier character.

        Args:
            char: Character to check

        Returns:
            True if valid identifier character
        """
        if not char:
            return False
        return bool(re.match(r'[a-zA-Z0-9_]', char))

    def is_digit(self, char: Optional[str]) -> bool:
        """
        Check if a character is a digit.

        Args:
            char: Character to check

        Returns:
            True if digit
        """
        if not char:
            return False
        return bool(re.match(r'[0-9]', char))

    def tokenize(self) -> List[Token]:
        """
        Tokenize the entire input.

        Returns:
            Array of tokens
        """
        self.tokens = []

        while not self.is_at_end():
            self.tokenize_next()

        # Add EOF token
        self.tokens.append(Token.eof(self.line, self.column))

        return self.tokens

    def tokenize_next(self):
        """Tokenize the next token from current position."""
        char = self.peek()
        start_line = self.line
        start_column = self.column

        # Newline
        if char in ['\n', '\r']:
            self.tokenize_newline()
            return

        # Whitespace (indentation at start of line)
        if char in [' ', '\t']:
            # Only create INDENT token at start of line
            if (self.column == 1 or len(self.tokens) == 0 or
                    self.tokens[-1].type == TokenType.NEWLINE):
                spaces = self.skip_whitespace()
                if spaces > 0:
                    self.tokens.append(Token.indent(spaces, start_line, start_column))
            else:
                # Skip inline whitespace
                self.skip_whitespace()
            return

        # Comment
        if char == '#':
            self.tokenize_comment()
            return

        # Section marker
        if char == '@':
            self.tokenize_section()
            return

        # Colon
        if char == ':':
            self.advance()
            self.tokens.append(Token(TokenType.COLON, ':', start_line, start_column, 1))
            return

        # Pipe
        if char == '|':
            self.advance()
            self.tokens.append(Token(TokenType.PIPE, '|', start_line, start_column, 1))
            return

        # Number (including negative) - Check BEFORE dash to handle negative numbers
        if self.is_digit(char) or (char == '-' and self.is_digit(self.peek_at(1))):
            self.tokenize_number()
            return

        # Dash (array item) - Only if not followed by digit (negative numbers handled above)
        if char == '-':
            self.advance()
            self.tokens.append(Token(TokenType.DASH, '-', start_line, start_column, 1))
            return

        # Left brace
        if char == '{':
            self.advance()
            self.tokens.append(Token(TokenType.LBRACE, '{', start_line, start_column, 1))
            return

        # Right brace
        if char == '}':
            self.advance()
            self.tokens.append(Token(TokenType.RBRACE, '}', start_line, start_column, 1))
            return

        # Left bracket
        if char == '[':
            self.advance()
            self.tokens.append(Token(TokenType.LBRACKET, '[', start_line, start_column, 1))
            return

        # Right bracket
        if char == ']':
            self.advance()
            self.tokens.append(Token(TokenType.RBRACKET, ']', start_line, start_column, 1))
            return

        # Comma
        if char == ',':
            self.advance()
            self.tokens.append(Token(TokenType.COMMA, ',', start_line, start_column, 1))
            return

        # Dot
        if char == '.':
            self.advance()
            self.tokens.append(Token(TokenType.DOT, '.', start_line, start_column, 1))
            return

        # Variable reference ($var)
        if char == '$':
            self.tokenize_reference('$', TokenType.VAR_REF)
            return

        # Object reference (&obj)
        if char == '&':
            self.tokenize_reference('&', TokenType.OBJ_REF)
            return

        # Quoted string
        if char in ['"', "'"]:
            self.tokenize_quoted_string(char)
            return

        # Identifier, keyword, or unquoted string
        if self.is_identifier_start(char):
            self.tokenize_identifier_or_keyword()
            return

        # Unknown character - skip it
        self.advance()

    def tokenize_newline(self):
        """Tokenize a newline (handles \\n, \\r\\n, \\r)."""
        start_line = self.line
        start_column = self.column
        value = ''

        if self.peek() == '\r' and self.peek_at(1) == '\n':
            value = self.advance() + self.advance()
        else:
            value = self.advance()

        self.tokens.append(Token(TokenType.NEWLINE, value, start_line, start_column, len(value)))

    def tokenize_comment(self):
        """Tokenize a comment (# or #| |#)."""
        start_line = self.line
        start_column = self.column

        self.advance()  # consume #

        # Multi-line comment start #|
        if self.peek() == '|':
            self.advance()  # consume |
            self.tokens.append(Token(TokenType.COMMENT_START, '#|', start_line, start_column, 2))
            return

        # Single-line comment - consume until newline
        value = '#'
        while not self.is_at_end() and self.peek() not in ['\n', '\r']:
            value += self.advance()

        self.tokens.append(Token(TokenType.COMMENT, value, start_line, start_column, len(value)))

    def tokenize_section(self):
        """Tokenize a section marker (@section_name)."""
        start_line = self.line
        start_column = self.column

        self.advance()  # consume @

        # Read section name (identifier)
        name = ''
        while self.is_identifier_char(self.peek()) or self.peek() == '.':
            name += self.advance()

        value = '@' + name
        self.tokens.append(Token(TokenType.SECTION, value, start_line, start_column, len(value)))

    def tokenize_reference(self, prefix: str, token_type: TokenType):
        """
        Tokenize a reference ($var or &obj).

        Args:
            prefix: Reference prefix ($ or &)
            token_type: Token type
        """
        start_line = self.line
        start_column = self.column

        self.advance()  # consume prefix

        # Check for keywords like $def: or $data:
        if prefix == '$':
            remaining = self.input[self.pos:self.pos + 5]
            if remaining.startswith('def:'):
                for _ in range(4):
                    self.advance()  # consume 'def:'
                self.tokens.append(Token(TokenType.DEF_KEYWORD, '$def:', start_line, start_column, 5))
                return
            if remaining.startswith('data:'):
                for _ in range(5):
                    self.advance()  # consume 'data:'
                self.tokens.append(Token(TokenType.DATA_KEYWORD, '$data:', start_line, start_column, 6))
                return

        # Read reference name
        name = ''
        while self.is_identifier_char(self.peek()):
            name += self.advance()

        value = prefix + name
        self.tokens.append(Token(token_type, value, start_line, start_column, len(value)))

    def tokenize_quoted_string(self, quote: str):
        """
        Tokenize a quoted string (supports " and ').

        Args:
            quote: Quote character (" or ')
        """
        start_line = self.line
        start_column = self.column

        self.advance()  # consume opening quote

        value = ''
        escaped = False

        while not self.is_at_end():
            char = self.peek()

            if escaped:
                value += self.advance()
                escaped = False
            elif char == '\\':
                value += self.advance()
                escaped = True
            elif char == quote:
                self.advance()  # consume closing quote
                break
            else:
                value += self.advance()

        # Return the value WITH quotes so parser can distinguish quoted vs unquoted
        full_value = quote + value + quote
        self.tokens.append(Token(TokenType.STRING, full_value, start_line, start_column, len(full_value)))

    def tokenize_number(self):
        """Tokenize a number (integer or float, including negative)."""
        start_line = self.line
        start_column = self.column

        value = ''

        # Handle negative sign
        if self.peek() == '-':
            value += self.advance()

        # Read digits
        while self.is_digit(self.peek()):
            value += self.advance()

        # Read decimal part
        if self.peek() == '.' and self.is_digit(self.peek_at(1)):
            value += self.advance()  # consume .
            while self.is_digit(self.peek()):
                value += self.advance()

        # Read scientific notation (e.g., 1.5e10, 2e-3)
        if self.peek() in ['e', 'E']:
            next_char = self.peek_at(1)
            has_sign = next_char in ['+', '-']
            char_after_sign = self.peek_at(2) if has_sign else next_char

            # Only treat as scientific notation if there's a digit after e/E (and optional sign)
            if self.is_digit(char_after_sign):
                value += self.advance()  # consume 'e' or 'E'
                if has_sign:
                    value += self.advance()  # consume '+' or '-'
                while self.is_digit(self.peek()):
                    value += self.advance()

        self.tokens.append(Token(TokenType.NUMBER, value, start_line, start_column, len(value)))

    def tokenize_identifier_or_keyword(self):
        """Tokenize an identifier, keyword, or unquoted string value."""
        start_line = self.line
        start_column = self.column

        value = ''

        # Read identifier characters
        while self.is_identifier_char(self.peek()) or self.peek() == '.':
            value += self.advance()

        # Check for keywords (true, false, null)
        if value in ['true', 'false']:
            self.tokens.append(Token(TokenType.BOOLEAN, value, start_line, start_column, len(value)))
            return

        if value == 'null':
            self.tokens.append(Token(TokenType.NULL, value, start_line, start_column, len(value)))
            return

        # Otherwise it's an identifier or unquoted string
        self.tokens.append(Token(TokenType.IDENTIFIER, value, start_line, start_column, len(value)))

    @staticmethod
    def filter_ignorable(tokens: List[Token]) -> List[Token]:
        """
        Filter out ignorable tokens (comments, whitespace).

        Args:
            tokens: Tokens to filter

        Returns:
            Filtered tokens
        """
        return [
            token for token in tokens
            if token.type not in [TokenType.COMMENT, TokenType.COMMENT_START, TokenType.COMMENT_END]
        ]
